Tests/Basic/SynchronizedQueueTests.swift

```diff
@@ -0,0 +1,143 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import Utility
+import Basic
+
+class SyncronizedQueueTests: XCTestCase {
+
+    func testSingleProducerConsumer() {
+        let queue = SynchronizedQueue<Int?>()
+        let queueElements = Set(0..<10)
+        var consumed = Set<Int>()
+
+        let producer = Thread {
+            for element in queueElements {
+                queue.enqueue(element)
+            }
+            queue.enqueue(nil)
+        }
+
+        let consumer = Thread {
+            while let element = queue.dequeue() {
+                consumed.insert(element)
+            }
+        }
+
+        consumer.start()
+        producer.start()
+        consumer.join()
+        producer.join()
+
+       XCTAssertEqual(consumed, queueElements)
+    }
+
+    func testMultipleProducerConsumer() {
+        let queue = SynchronizedQueue<Int?>()
+
+        let queueElementsOne = Set(0..<100)
+        let queueElementsTwo = Set(100..<500)
+
+        var consumed = Set<Int>()
+        var consumedLock = Basic.Lock()
+
+        // Create two producers.
+        let producers = [queueElementsOne, queueElementsTwo].map { queueElements in
+            return Thread {
+                for element in queueElements {
+                    queue.enqueue(element)
+                }
+                queue.enqueue(nil)
+            }
+        }
+
+        // Create two consumers.
+        let consumers = [0, 1].map { _ in
+            return Thread {
+                while let element = queue.dequeue() {
+                    consumedLock.withLock {
+                        consumed.insert(element)
+                    }
+                }
+            }
+        }
+
+        consumers.forEach { $0.start() }
+        producers.forEach { $0.start() }
+        // Block until all producers and consumers are done.
+        consumers.forEach { $0.join() }
+        producers.forEach { $0.join() }
+
+        // Make sure everything was consumed.
+        XCTAssertEqual(consumed, queueElementsOne.union(queueElementsTwo))
+    }
+
+
+    // Stress test for queue. Can produce an element only when current element gets consumed.
+    func testMultipleProducerConsumer2() {
```

Comment:
It might be nice to explain what this is stress testing (the blocking/signaling of dequeue, because the expectation is that often the client will see an empty queue).